### PR TITLE
Turf fire

### PR DESCRIPTION
**Pull request by rye-rice**
## About The Pull Request

https://github.com/shiptest-ss13/Shiptest/pull/848

Now includes a special resprite by electricpants#6824 for all prebuilt flamethrowers!
image
image

## Why It's Good For The Game
This also makes flamethrowers less of a grief tool, as now they can work like a AOE denial weapon, and are easily countered by extingishers

## Changelog
:cl:
add: Turf fires, again
/:cl:

### CODE REVIEW

**Comment by rye-rice** [comment]
This should be ready for review again

**Comment by rye-rice** [comment]
How do you make stateless go away

**Comment by ZephyrTFA** [comment]
Stale doesn't go away until a maintainer performs some interaction with your PR, such as leaving a comment or review

**Comment by MarkSuckerberg** [comment]
a new commit will remove it too

File: code/game/objects/effects/turf_fire.dm [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+	SSturf_fire.fires -= src
+	return ..()
+
+/obj/effect/abstract/turf_fire/proc/process_waste()
+	var/turf/open/open_turf = loc
+	if(open_turf.planetary_atmos)
+		return TRUE
+	var/datum/gas_mixture/cached_air = open_turf.air
+	var/oxy = cached_air.get_moles(GAS_O2)
+	if (oxy < 0.5)
```
Comment (Fikou) [inline code/game/objects/effects/turf_fire.dm]: use a define

File: code/game/objects/effects/turf_fire.dm [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+		if(fire_power <= 0)
+			qdel(src)
+			return
+	open_turf.hotspot_expose(TURF_FIRE_TEMP_BASE + (TURF_FIRE_TEMP_INCREMENT_PER_POWER*fire_power), TURF_FIRE_VOLUME)
+	for(var/atom/movable/burning_atom as anything in open_turf)
+		burning_atom.fire_act(TURF_FIRE_TEMP_BASE + (TURF_FIRE_TEMP_INCREMENT_PER_POWER*fire_power), TURF_FIRE_VOLUME)
+	if(interact_with_atmos)
+		if(prob(fire_power))
+			open_turf.burn_tile()
+		if(prob(6))
```
Comment (Fikou) [inline code/game/objects/effects/turf_fire.dm]: define/variable here too

File: code/game/objects/items/flamethrower.dm [hunk 3]
```diff
@@ -230,28 +238,94 @@
...
+	if(!linked_flamethrower)
+		return
+	if(!linked_flamethrower.beaker)
+		return
+	var/turf/location = get_turf(src)
+
+	var/datum/reagents/beaker_reagents = linked_flamethrower.beaker.reagents
+	var/datum/reagents/my_fraction = new()
+
+	beaker_reagents.trans_to(my_fraction, release_amount, no_react = TRUE)
```
Comment (Fikou) [inline code/game/objects/items/flamethrower.dm]: don't link the flamethrower to it, just transfer the reagents when the projectile is first created by the flamethrower

File: code/game/turfs/change_turf.dm [hunk 4]
```diff
@@ -157,17 +157,26 @@ GLOBAL_LIST_INIT(blacklisted_automated_baseturfs, typecacheof(list(
 
 /turf/open/ChangeTurf(path, list/new_baseturfs, flags) //Resist the temptation to make this default to keeping air.
 	if ((flags & CHANGETURF_INHERIT_AIR) && ispath(path, /turf/open))
+		var/turf_fire_ref
+		if(turf_fire)
+			if(ispath(path, /turf/open/openspace) || ispath(path, /turf/open/space))
+				qdel(turf_fire)
+			else
+				turf_fire_ref = turf_fire
```
Comment (Fikou) [inline code/game/turfs/change_turf.dm]: wouldnt it be better to run this after we are sure whether it succeeded/failed?

File: code/game/objects/effects/turf_fire.dm [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+/obj/effect/abstract/turf_fire/Initialize(mapload, power, fire_color)
+	. = ..()
+	var/turf/open/open_turf = loc
+	if(open_turf.turf_fire)
+		return INITIALIZE_HINT_QDEL
+
+	var/static/list/loc_connections = list(
+		COMSIG_ATOM_ENTERED = .proc/on_entered,
+	)
+	AddElement(/datum/element/connect_loc, src, loc_connections)
```
Comment (Fikou) [inline code/game/objects/effects/turf_fire.dm]: there isnt a second argument to connect_loc, the src shouldnt be there

**Review by Fikou — CHANGES_REQUESTED** [review]
code looks good otherwise, nice job

File: code/game/turfs/open/_open.dm [hunk 6]
```diff
@@ -261,3 +263,11 @@
 		air.set_moles(GAS_CO2, max(air.get_moles(GAS_CO2)-(pulse_strength/1000),0))
 		air.set_moles(GAS_O2, max(air.get_moles(GAS_O2)-(pulse_strength/2000),0))
 		air.adjust_moles(GAS_PLUOXIUM, pulse_strength/4000)
+
+/turf/open/IgniteTurf(power, fire_color)
+	if(turf_fire)
+		turf_fire.AddPower(power)
+		return
+	if(isopenspaceturf(src) || isspaceturf(src))
```
Comment (Fikou) [inline code/game/turfs/open/_open.dm]: use isgroundlessturf, you probably dont need isopenspaceturf defined if you do that

File: code/modules/cargo/packs.dm [hunk 7]
```diff
@@ -630,12 +630,11 @@
 
 /datum/supply_pack/security/armory/fire
 	name = "Incendiary Weapons Crate"
-	desc = "Burn, baby burn. Contains three incendiary grenades, three plasma canisters, and a flamethrower. Requires Armory access to open."
+	desc = "Burn, baby burn. Contains three incendiary grenades and three prebuilt flamethrowers. Requires Armory access to open."
```
Comment (Fikou) [inline code/modules/cargo/packs.dm:633]:
```suggestion
	desc = "Burn, baby burn. Contains three incendiary grenades and two prebuilt flamethrowers. Requires Armory access to open."
```

File: code/modules/cargo/packs.dm [hunk 8]
```diff
@@ -630,12 +630,11 @@
...
 	cost = 4500
 	contains = list(/obj/item/flamethrower/full,
```
Comment (Fikou) [inline code/modules/cargo/packs.dm]:
```suggestion
	contains = list(
		/obj/item/flamethrower/full,
```

Comment (Fikou) [reply code/modules/cargo/packs.dm]: also lower the indentation of the other ones please

File: code/game/turfs/open/_open.dm (hunk 6 above)
Comment (rye-rice) [reply code/game/turfs/open/_open.dm]: isgroundlessturf

File: code/game/turfs/change_turf.dm (hunk 4 above)
Comment (rye-rice) [reply code/game/turfs/change_turf.dm]: moved this after, added a isgroundlessturf, and seems to work moved

**Comment by rye-rice** [comment]
suffering

File: code/modules/cargo/packs.dm [hunk 9]
```diff
@@ -630,12 +630,12 @@
...
-	desc = "Burn, baby burn. Contains three incendiary grenades, three plasma canisters, and a flamethrower. Requires Armory access to open."
+	desc = "Burn, baby burn. Contains three incendiary grenades and three prebuilt flamethrowers. Requires Armory access to open."
 	cost = 4500
-	contains = list(/obj/item/flamethrower/full,
-					/obj/item/tank/internals/plasma,
-					/obj/item/tank/internals/plasma,
-					/obj/item/tank/internals/plasma,
+	contains = list(
+		/obj/item/flamethrower/full,
+					/obj/item/flamethrower/full,
```
Comment (Fikou) [inline code/modules/cargo/packs.dm]: u didnt move the indentation of the other ones down

File: code/game/turfs/change_turf.dm [hunk 10]
```diff
@@ -164,10 +164,19 @@ GLOBAL_LIST_INIT(blacklisted_automated_baseturfs, typecacheof(list(
 			QDEL_NULL(stashed_air)
 			return
 		var/turf/open/newTurf = .
+		var/turf_fire_ref
+		if(turf_fire)
+			if(isgroundlessturf(path))
```
Comment (Fikou) [inline code/game/turfs/change_turf.dm]: isgroundlessturf is for instances, not paths. you dont need to use paths after the turf has been created

File: code/modules/cargo/packs.dm (hunk 9 above)
Comment (rye-rice) [reply code/modules/cargo/packs.dm]: fixed, i am the stupid

File: code/game/turfs/change_turf.dm (hunk 10 above)
Comment (rye-rice) [reply code/game/turfs/change_turf.dm]: i reverted this to how it was before

File: code/modules/cargo/packs.dm [hunk 11]
```diff
@@ -630,15 +630,16 @@
...
-					/obj/item/grenade/chem_grenade/incendiary,
-					/obj/item/grenade/chem_grenade/incendiary)
+	contains = list(
+		/obj/item/flamethrower/full,
+		/obj/item/flamethrower/full,
+		/obj/item/flamethrower/full,
+		/obj/item/grenade/chem_grenade/incendiary,
+		/obj/item/grenade/chem_grenade/incendiary,
+		/obj/item/grenade/chem_grenade/incendiary,
+		)
```
Comment (Fikou) [inline code/modules/cargo/packs.dm]:
```suggestion
	)
```

File: code/game/turfs/change_turf.dm [hunk 12]
```diff
@@ -164,10 +164,19 @@ GLOBAL_LIST_INIT(blacklisted_automated_baseturfs, typecacheof(list(
 			QDEL_NULL(stashed_air)
 			return
 		var/turf/open/newTurf = .
+		var/turf_fire_ref
+		if(turf_fire)
+			if(ispath(path, /turf/open/openspace) || ispath(path, /turf/open/space))
```
Comment (Fikou) [inline code/game/turfs/change_turf.dm]: again, you dont need to operate on paths. you already created the turf.

Comment (rye-rice) [reply code/game/turfs/change_turf.dm]: Agony, suffering even

File: code/game/turfs/change_turf.dm (hunk 10 above)
Comment (rye-rice) [reply code/game/turfs/change_turf.dm]: see below

**Comment by Fikou** [comment]
let's go?

**Review by Fikou — APPROVED** [review]

**Pull request merged**